tutorials/identifying-first-day-no-snow/index.md

```diff
@@ -0,0 +1,516 @@
+}
+```
+
+### 3\. Define an analysis mask
+
+Here is the opportunity to define a mask for your analysis. This mask can be
+used to constrain the analysis to certain latitudes
+(`ee.Image.pixelLonLat()`), land cover types, geometries, etc. In this case
+we will: 1) mask out water so that the analysis is confined to pixels over
+landforms only; 2) mask out pixels that have very few days of snow cover; 3)
+mask out pixels that are snow covered for a good deal of the year
+(e.g., glaciers).
+
+Import the MODIS water/land mask dataset, select the ‘water_mask’ band, and
+set all land pixels to value 1.
+
+```js
+var waterMask = ee.Image('MODIS/MOD44W/MOD44W_005_2000_02_24')
+  .select('water_mask')
+  .not();
+```
+
+Import the MODIS Snow Cover Daily Global 500m product and select the
+'NDSI_Snow_Cover' band.
+
+**Note that this dataset is used in following
+sections. If you decide not to do any analysis masking, you must still
+include and run this line before the start of the next section**.
+
+```js
+var completeCol = ee.ImageCollection('MODIS/006/MOD10A1')
+  .select('NDSI_Snow_Cover');
+```
+
+Mask pixels based on frequency of snow cover.
+
+```js
+// Pixels must have been 10% snow covered for at least 2 weeks in 2018.
+var snowCoverEphem = completeCol.filterDate('2018-01-01', '2019-01-01')
+  .map(function(img) {
+    return img.gte(10);
+  })
+  .sum()
+  .gte(14);
+
+// Pixels must not be 10% snow covered more than 124 days in 2018.
+var snowCoverConst = completeCol.filterDate('2018-01-01', '2019-01-01')
+  .map(function(img) {
+    return img.gte(10);
+  })
+  .sum()
+  .lt(125);
+```
+
+Combine the water mask and the snow cover frequency masks.
+
+```js
+var analysisMask = waterMask.multiply(snowCoverEphem).multiply(snowCoverConst);
+```
+
+### 4\. Identify the first day of the year without snow per pixel, per year
+
+Make a list of the years to process; input variables were defined in step 1.
+
+```js
+var years = ee.List.sequence(startYear, endYear);
+```
+
+Map the following function over the list of years. For each year, identify
+the first day with zero percent snow cover.
+ 
+1. Define the start and end dates to filter the dataset for the given year.
+2. Filter the image collection by the date range.
+3. Add the date bands to each image in the filtered collection.
+4. Sort the filtered collection by date.
+5. Make a mosaic using the `min` reducer to select the pixel with 0 (minimum)
+snow cover. Since the collection is sorted by date, the first image with 0
+snow cover is selected. This operation is conducted per-pixel to build the
+complete image mosaic.
+6. Apply the analysis mask to the resulting mosaic.
+ 
+An `ee.List` of images is returned.
+
+```js
+var annualList = years.map(function(year) {
+  // Set the global startYear variable as the year being worked on so that
+  // it will be accessible to the addDateBands mapped to the collection below.
+  startYear = year;
```

Comment:
Could we make `startYear` and `startDate` local variables and pass them into `addDateBands` instead of having them be global variables? It may be better to do this to set a good example for others, since with larger projects, global vars can be harder to manage, leading to bugs in client code (e.g., in cases when global state can be changed at multiple points in code, the flow of information is harder to understand and manage).

Reply:
`addDateBands` is a function that gets applied to a collection of images using `.map()`. Functions applied with `.map()` can only accept a single argument: the image.

Reply:
No, you're right, `var` scope isn't hierarchical as I assume, it would require more advanced JS usage to make this work, so please disregard!

Reply:
Ack